src/server/zset_family.cc

```diff
@@ -1351,6 +1351,74 @@ void ZSetFamily::ZCount(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+vector<ScoredMap> OpDiff(EngineShard* shard, Transaction* t) {
+  ArgSlice keys = t->GetShardArgs(shard->shard_id());
+  DVLOG(1) << "shard:" << shard->shard_id() << ", keys " << vector(keys.begin(), keys.end());
+  DCHECK(!keys.empty());
+
+  vector<ScoredMap> results;
+  results.reserve(keys.size());
+
+  auto& db_slice = shard->db_slice();
+  for (size_t i = 0; i < keys.size(); ++i) {
+    auto it = db_slice.Find(t->GetDbContext(), keys[i], OBJ_ZSET);
+    if (!it) {
+      results.push_back({});
+      continue;
+    }
+
+    ScoredMap sm = FromObject((*it)->second, 1);
+    results.push_back(std::move(sm));
+  }
+
+  return results;
+}
+
+void ZSetFamily::ZDiff(CmdArgList args, ConnectionContext* cntx) {
+  vector<vector<ScoredMap>> maps(shard_set->size());
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    maps[shard->shard_id()] = OpDiff(shard, t);
+    return OpStatus::OK;
+  };
+
+  cntx->transaction->ScheduleSingleHop(std::move(cb));
+
+  const string_view key = ArgS(args, 1);
+  const ShardId sid = Shard(key, maps.size());
+  auto& sm = maps[sid];
+  if (sm.empty()) {
+    (*cntx)->SendEmptyArray();
+    return;
+  }
+  auto result = std::move(sm[0]);
+  sm.erase(sm.begin());
+  for (auto& vsm : maps) {
+    for (auto& sm : vsm) {
+      for (auto& [key, value] : sm) {
+        auto it = result.find(key);
+        if (it != result.end()) {
+          result.erase(it);
+        }
+      }
+    }
+  }
```

Comment:
It achieves a simple result, but is hard to read because of the short variable names and branched loops. Alternatively, add two comments to it like the following: 1. Extract map of first key 2. Filter out all elements of the remaining sets

Reply:
I added comments and a small touch ;)